Sources/AsyncHTTPClient/ConnectionPool/ChannelHandler/HTTP1ProxyConnectHandler.swift

```diff
@@ -0,0 +1,204 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+final class HTTP1ProxyConnectHandler: ChannelDuplexHandler, RemovableChannelHandler {
+    typealias OutboundIn = Never
+    typealias OutboundOut = HTTPClientRequestPart
+    typealias InboundIn = HTTPClientResponsePart
+
+    enum State {
+        // transitions to `.connectSent` or `.failed`
+        case initialized
+        // transitions to `.headReceived` or `.failed`
+        case connectSent(Scheduled<Void>)
+        // transitions to `.completed` or `.failed`
+        case headReceived(Scheduled<Void>)
+        // final error state
+        case failed(Error)
+        // final success state
+        case completed
+    }
+
+    private var state: State = .initialized
+
+    let targetHost: String
+    let targetPort: Int
+    let proxyAuthorization: HTTPClient.Authorization?
+    let deadline: NIODeadline
+
+    private var proxyEstablishedPromise: EventLoopPromise<Void>?
+    var proxyEstablishedFuture: EventLoopFuture<Void>? {
+        return self.proxyEstablishedPromise?.futureResult
+    }
+
+    init(targetHost: String,
+         targetPort: Int,
+         proxyAuthorization: HTTPClient.Authorization?,
+         deadline: NIODeadline) {
+        self.targetHost = targetHost
+        self.targetPort = targetPort
+        self.proxyAuthorization = proxyAuthorization
+        self.deadline = deadline
+    }
+
+    func handlerAdded(context: ChannelHandlerContext) {
+        self.proxyEstablishedPromise = context.eventLoop.makePromise(of: Void.self)
+
+        if context.channel.isActive {
+            self.sendConnect(context: context)
```

Comment:
If the channel is already active, won't `channelActive` have already been called, meaning `sendConnect` will have been called? If that's the case, this whole `if` is redundant as it should probably be guarded in the `sendConnect` function.

Reply:
No. This is `handlerAdded`: if the channel is active, we're being added _after_ `channelActive` fired, and so we'll never see it.

Reply:
Yeh, I'm saying we can remove the `if` because it will always be true.

Reply:
It won't always be true: we might be added _before_ `channelActive` is called, in which case this will be `false`.

Nonetheless I agree that the state machine is already policing this.